File: sht1/argonfluid/fortran/zmclj.f90
!**********************************************************************!
!
! File: zmclj.f90
!
! NVT-Monte Carlo of Lennard-Jonesium
! Re-initialize checkpoint file
!
! 13-Apr-1997 (MN)
! 17-Apr-2012
!
!**********************************************************************!

module getval_m

!**********************************************************************!

! Generic subroutine: read item from standard input/keep old value on
! hitting return

  implicit none

  private::getint,getreal,getstrg
  public::getval

  interface getval
    module procedure getint,getreal,getstrg
  end interface

  character(len=80),private::line

contains

!**********************************************************************!

  subroutine getint(intval)

!**********************************************************************!

    integer,intent(in out)::intval

    read(unit=*,fmt="(a)") line
    if(len_trim(adjustl(line))>0) then
      read(unit=line,fmt=*) intval
    end if

    return

  end subroutine getint

!**********************************************************************!

  subroutine getreal(realval)

!**********************************************************************!

    real,intent(in out)::realval

    read(unit=*,fmt="(a)") line
    if(len_trim(adjustl(line))>0) then
      read(unit=line,fmt=*) realval
    end if

    return

  end subroutine getreal

!**********************************************************************!

  subroutine getstrg(strgval)

!**********************************************************************!

    character(len=*),intent(in out)::strgval

    read(unit=*,fmt="(a)") line
    if(len_trim(adjustl(line))>0) then
      read(unit=line,fmt="(a)") strgval
    end if

    return

  end subroutine getstrg

end module getval_m

!**********************************************************************!

program zmclj

!**********************************************************************!

  use getval_m

  implicit none

  integer,parameter::iin=1,iout=2

  character(len=80)::fname
  integer::n,nran,nt,ntjob,ntprint,ntskip
  integer,dimension(:),allocatable::iran
  real::disp,dr,fact,rho,rhon,t
  real,dimension(:),allocatable::x,y,z

! RNG characteristics

  call random_seed(size=nran)

  allocate(iran(nran))

! Read old checkpoint file

  fname="mclj_out.dat"
  write(unit=*,fmt="(a)",advance="no") "         infile=[mclj_out.dat] "
  call getval(fname)

  open(unit=iin,file=fname,status="old",action="read", &
       form="unformatted",position="rewind")

  read(unit=iin) n,nt,ntjob,ntprint,ntskip,iran
  read(unit=iin) disp,dr,rho,t

! Allocate arrays

  allocate(x(n),y(n),z(n))

! Positions

  read(unit=iin) x,y,z

  close(unit=iin)

  rhon=rho

! User input

  write(unit=*,fmt="(a,i13)") "              n=",n
  write(unit=*,fmt="(a,f12.5,a)",advance="no") "            rho=[",rhon,"] "
  call getval(rhon)
  write(unit=*,fmt="(a,f12.5,a)",advance="no") "              t=[",t,"] "
  call getval(t)
  write(unit=*,fmt="(a,f12.5,a)",advance="no") "           disp=[",disp,"] "
  call getval(disp)
  write(unit=*,fmt="(a,f12.5,a)",advance="no") "             dr=[",dr,"] "
  call getval(dr)
  write(unit=*,fmt="(a,i12,a)",advance="no") "         ntskip=[",ntskip,"] "
  call getval(ntskip)
  write(unit=*,fmt="(a,i12,a)",advance="no") " ntprint/ntskip=[",ntprint,"] "
  call getval(ntprint)
  write(unit=*,fmt="(a,i12,a)",advance="no") "   ntjob/ntskip=[",ntjob,"] "
  call getval(ntjob)

! Rescale positions

  if(rhon/=rho) then
    fact=(rho/rhon)**(1.0/3.0)
    rho=rhon
    x=fact*x
    y=fact*y
    z=fact*z
  end if

! Write new startup file

  fname="mclj_in.dat"
  write(unit=*,fmt="(a)",advance="no") "        outfile=[ mclj_in.dat] "
  call getval(fname)

  nt=0

  open(unit=iout,file=fname,status="replace",action="write", &
       form="unformatted")

  write(unit=iout) n,nt,ntjob,ntprint,ntskip,iran
  write(unit=iout) disp,dr,rho,t
  write(unit=iout) x,y,z

  close(unit=iout)

! Deallocate arrays

  deallocate(iran,x,y,z)

end program zmclj

!**********************************************************************!

File: lj-canonical/zmclj.c
/**********************************************************************
 *
 * File: zmclj.c
 *
 * NVT-Monte Carlo of Lennard-Jonesium
 * Re-initialize checkpoint file
 *
 * 12-May-2007 (MN)
 * 17-Apr-2012
 *
 **********************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

#include "getval.h"

#define NL '\n'
#define BSIZE 80

/**********************************************************************/

int main() {

/**********************************************************************/

  FILE *fpi,*fpo;
  char fname[BSIZE];
  unsigned short seed[3];
  int n,nt,ntjob,ntprint,ntskip;
  int i;
  double disp,dr,rho,t;
  double fact,rhon;
  double *x,*y,*z;

  /* Read old checkpoint file */

  strcpy(fname,"mclj_out.dat");
  printf("         infile=[mclj_out.dat] ");
  getval_s(fname);

  fpi=fopen(fname,"r");

  fread(&n,sizeof(int),1,fpi);
  fread(&nt,sizeof(int),1,fpi);
  fread(&ntjob,sizeof(int),1,fpi);
  fread(&ntprint,sizeof(int),1,fpi);
  fread(&ntskip,sizeof(int),1,fpi);
  fread(seed,sizeof(unsigned short),3,fpi);

  fread(&disp,sizeof(double),1,fpi);
  fread(&dr,sizeof(double),1,fpi);
  fread(&rho,sizeof(double),1,fpi);
  fread(&t,sizeof(double),1,fpi);

  /* Allocate arrays */

  x=(double*)malloc(n*sizeof(double));
  y=(double*)malloc(n*sizeof(double));
  z=(double*)malloc(n*sizeof(double));

  /* Positions */

  fread(x,sizeof(double),n,fpi);
  fread(y,sizeof(double),n,fpi);
  fread(z,sizeof(double),n,fpi);

  fclose(fpi);

  rhon=rho;

  /* User input */

  printf("              n=%13d\n",n);
  printf("            rho=[%12.5lf] ",rho);
  getval_d(&rhon);
  printf("              t=[%12.5lf] ",t);
  getval_d(&t);
  printf("           disp=[%12.5lf] ",disp);
  getval_d(&disp);
  printf("             dr=[%12.5lf] ",dr);
  getval_d(&dr);
  printf("         ntskip=[%12d] ",ntskip);
  getval_i(&ntskip);
  printf(" ntprint/ntskip=[%12d] ",ntprint);
  getval_i(&ntprint);
  printf("   ntjob/ntskip=[%12d] ",ntjob);
  getval_i(&ntjob);

  /* Rescale positions */

  if(rhon!=rho) {
    fact=cbrt(rho/rhon);
    rho=rhon;
    for(i=0;i<=n-1;i++) {
      x[i]=fact*x[i];
      y[i]=fact*y[i];
      z[i]=fact*z[i];
    }
  }

  /* Write new startup file */

  strcpy(fname,"mclj_in.dat");
  printf("        outfile=[ mclj_in.dat] ");
  getval_s(fname);

  nt=0;

  fpo=fopen(fname,"w");

  fwrite(&n,sizeof(int),1,fpo);
  fwrite(&nt,sizeof(int),1,fpo);
  fwrite(&ntjob,sizeof(int),1,fpo);
  fwrite(&ntprint,sizeof(int),1,fpo);
  fwrite(&ntskip,sizeof(int),1,fpo);
  fwrite(seed,sizeof(unsigned short),3,fpo);

  fwrite(&disp,sizeof(double),1,fpo);
  fwrite(&dr,sizeof(double),1,fpo);
  fwrite(&rho,sizeof(double),1,fpo);
  fwrite(&t,sizeof(double),1,fpo);

  fwrite(x,sizeof(double),n,fpo);
  fwrite(y,sizeof(double),n,fpo);
  fwrite(z,sizeof(double),n,fpo);

  fclose(fpo);

  /* Deallocate arrays */

  free(x);
  free(y);
  free(z);

  return 0;

}

/**********************************************************************/
